### PR TITLE
Feature/mempool integration

### PR DESCRIPTION
**Pull request by Mr-Leshiy**

### CODE REVIEW

File: src/validation.cpp [hunk 1]
```diff
@@ -2626,6 +2609,9 @@ bool CChainState::ConnectTip(BlockValidationState& state, const CChainParams& ch
     mempool.removeForBlock(blockConnecting.vtx, pindexNew->nHeight);
     disconnectpool.removeForBlock(blockConnecting.vtx);
 
+    // VeriBlock: remove from pop_mempool
+    VeriBlock::getService<VeriBlock::PopService>().removePayloads(blockConnecting.v_popData);
```
Comment (Warchant) [inline src/validation.cpp]:
```suggestion
    VeriBlock::getService<VeriBlock::PopService>().removePayloadsFromMempool(blockConnecting.v_popData);
```

File: src/vbk/pop_service_impl.cpp [hunk 2]
```diff
@@ -48,6 +48,16 @@ void popstack(std::vector<valtype>& stack)
         throw std::runtime_error("popstack(): stack empty");
     stack.pop_back();
 }
+*/
+} // namespace
```
Comment (Warchant) [inline src/vbk/pop_service_impl.cpp]: you can remove that

File: src/vbk/rpc_register.cpp [hunk 3]
```diff
@@ -186,20 +154,37 @@ UniValue submitpop(const JSONRPCRequest& request)
...
     LogPrint(BCLog::POP, "VeriBlock-PoP: submitpop RPC called with 1 ATV and %d VTBs\n", vtb_array.size());
+    std::vector<altintegration::VTB> vtbs;
     for (uint32_t idx = 0u, size = vtb_array.size(); idx < size; ++idx) {
         auto& vtbhex = vtb_array[idx];
-        auto vtb = ParseHexV(vtbhex, "vtb[" + std::to_string(idx) + "]");
-        script << vtb << OP_CHECKVTB;
+        auto vtb_bytes = ParseHexV(vtbhex, "vtb[" + std::to_string(idx) + "]");
+        vtbs.push_back(altintegration::VTB::fromVbkEncoding(vtb_bytes));
+        
+        LogPrint(BCLog::POP, "VeriBlock-PoP: VTB%d=\"%s\"\n", idx, vtbhex.get_str());
```
Comment (Warchant) [inline src/vbk/rpc_register.cpp]: remove that log

File: src/vbk/rpc_register.cpp [hunk 4]
```diff
@@ -186,20 +154,37 @@ UniValue submitpop(const JSONRPCRequest& request)
 
     CScript script;
 
+    // TODO put atv and vtbs into the mempool
```
Comment (Warchant) [inline src/vbk/rpc_register.cpp]: is it still a TODO?

File: src/vbk/rpc_register.cpp [hunk 5]
```diff
@@ -186,20 +154,37 @@ UniValue submitpop(const JSONRPCRequest& request)
...
+        LogPrint(BCLog::POP, "VeriBlock-PoP: VTB%d=\"%s\"\n", idx, vtbhex.get_str());
     }
 
     auto& atvhex = request.params[0];
-    auto atv = ParseHexV(atvhex, "atv");
-    script << atv << OP_CHECKATV;
-    script << OP_CHECKPOP;
+    auto atv_bytes = ParseHexV(atvhex, "atv");
+
+    LogPrint(BCLog::POP, "VeriBlock-PoP: ATV=\"%s\"\n", atvhex.get_str());
```
Comment (Warchant) [inline src/vbk/rpc_register.cpp]: remove

File: src/vbk/rpc_register.cpp [hunk 6]
```diff
@@ -186,20 +154,37 @@ UniValue submitpop(const JSONRPCRequest& request)
...
+
+    auto& pop_service = VeriBlock::getService<VeriBlock::PopService>();
+    auto& pop_mempool = pop_service.getMemPool();
+
+    altintegration::ValidationState state;
+    if (pop_mempool.submitATV({ altintegration::ATV::fromVbkEncoding(atv_bytes) }, state)) {
```
Comment (Warchant) [inline src/vbk/rpc_register.cpp]: Maybe this?
```suggestion
    if (!pop_mempool.submitATV({ altintegration::ATV::fromVbkEncoding(atv_bytes) }, state)) {
```

File: src/vbk/rpc_register.cpp [hunk 7]
```diff
@@ -186,20 +154,37 @@ UniValue submitpop(const JSONRPCRequest& request)
...
+        LogPrint(BCLog::POP, "VeriBlock--invalid-PoP: %s ", state.GetPath());
+        return "ivalid ATV";
+    }
+    if (pop_mempool.submitVTB(vtbs, state)) {
```
Comment (Warchant) [inline src/vbk/rpc_register.cpp]: same

File: src/vbk/rpc_register.cpp [hunk 8]
```diff
@@ -186,20 +154,37 @@ UniValue submitpop(const JSONRPCRequest& request)
...
+        LogPrint(BCLog::POP, "VeriBlock--invalid-PoP: %s ", state.GetPath());
```
Comment (Warchant) [inline src/vbk/rpc_register.cpp]: prefix for logs should be `VeriBlock-PoP` (used by Grafana)

File: src/validation.cpp (hunk 1 above)
Comment (Mr-Leshiy) [reply src/validation.cpp]: fixed

File: src/vbk/rpc_register.cpp (hunk 4 above)
Comment (Mr-Leshiy) [reply src/vbk/rpc_register.cpp]: fixed

**Review by Warchant — APPROVED** [review]

**Pull request merged**